### PR TITLE
Add iosxr grpc plugin and grpc_nw_get module

### PR DESCRIPTION
**Pull request by ganeshrn**
##### SUMMARY

*  Add iosxrv 9k gRPC plugin implementation
*  Add `grpc_nw_get` module to fetch configuration
   and state data from the network device with gRPC enabled
*  Add integration test for `grpc_nw_get` module

Related to
1) https://github.com/ansible/ansible/pull/52823

2) https://github.com/ansible/ansible/issues/38820

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

lib/ansible/module_utils/network/grpc/__init__.py
lib/ansible/module_utils/network/grpc/grpc.py
lib/ansible/modules/network/grpc/__init__.py
lib/ansible/modules/network/grpc/grpc_nw_get.py
lib/ansible/plugins/grpc/__init__.py
lib/ansible/plugins/grpc/iosxr.py
lib/ansible/plugins/grpc/iosxr_pb/__init__.py
lib/ansible/plugins/grpc/iosxr_pb/ems_grpc_pb2.py

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: lib/ansible/constants.py
Comment (Qalthos) [inline lib/ansible/constants.py]: It doesn't look like `grpc` is actually configurable.

File: lib/ansible/plugins/grpc/__init__.py
Comment (Qalthos) [inline lib/ansible/plugins/grpc/__init__.py:26]:
```suggestion
    @abstractmethod
    def get_config(self, section=None):
```

File: lib/ansible/plugins/grpc/__init__.py
Comment (Qalthos) [inline lib/ansible/plugins/grpc/__init__.py:35]:
```suggestion
    @abstractmethod
    def get(self, section=None):
```

File: lib/ansible/plugins/grpc/__init__.py
Comment (Qalthos) [inline lib/ansible/plugins/grpc/__init__.py:44]:
```suggestion
    @abstractmethod
    def edit_config(self, config=None, action=None):
```

Comment (Qalthos) [reply lib/ansible/plugins/grpc/__init__.py:44]: Oh, but I see this one isn't implemented... is that for lack of support, or just because the module didn't call for it?

**Review by Qalthos — DISMISSED** [review]
My biggest concern here is whether we need a new `grpc` plugin type at all. It's hard to tell with just one implementation, but it takes no options and mostly takes direction from variables in `self._connection`. If the location of the pb file could be inferred from `network_os` (or whatever else), then I don't see a compelling reason to have a `grpc` plugin, though admittedly I have no idea how any of this translates to another platform (if at all)

File: lib/ansible/config/base.yml
Comment (samccann) [inline lib/ansible/config/base.yml]:
```suggestion
  description: Colon separated paths in which Ansible will search for gRPC plugins.
```

File: lib/ansible/config/base.yml
Comment (samccann) [inline lib/ansible/config/base.yml]:
```suggestion
  name: gRPC Plugins Path
```

File: lib/ansible/modules/network/grpc/grpc_nw_get.py
Comment (samccann) [inline lib/ansible/modules/network/grpc/grpc_nw_get.py]:
```suggestion
    - gRPC is a high performance, open-source universal RPC framework.
```

File: lib/ansible/modules/network/grpc/grpc_nw_get.py
Comment (samccann) [inline lib/ansible/modules/network/grpc/grpc_nw_get.py]:
```suggestion
        the data to be retrieved from the target host device. If this option is not specified the entire
```

File: lib/ansible/modules/network/grpc/grpc_nw_get.py
Comment (samccann) [inline lib/ansible/modules/network/grpc/grpc_nw_get.py]:
```suggestion
        in result. This option is supported if the gRPC target host supports executing the CLI command
```

File: lib/ansible/modules/network/grpc/grpc_nw_get.py
Comment (samccann) [inline lib/ansible/modules/network/grpc/grpc_nw_get.py]:
```suggestion
        over the gRPC connection.
```

File: lib/ansible/modules/network/grpc/grpc_nw_get.py
Comment (samccann) [inline lib/ansible/modules/network/grpc/grpc_nw_get.py]:
```suggestion
  - This module supports the use of connection=grpc.
```

File: lib/ansible/modules/network/grpc/grpc_nw_get.py
Comment (samccann) [inline lib/ansible/modules/network/grpc/grpc_nw_get.py]:
```suggestion
  - This module requires the value of 'ansible_network_os'  be defined as an inventory variable.
```

File: lib/ansible/modules/network/grpc/grpc_nw_get.py
Comment (samccann) [inline lib/ansible/modules/network/grpc/grpc_nw_get.py]:
```suggestion
               is a valid json string
```

File: lib/ansible/plugins/grpc/iosxr.py
Comment (samccann) [inline lib/ansible/plugins/grpc/iosxr.py]:
```suggestion
short_description: gRPC Plugin for IOS XR devices
```

File: lib/ansible/plugins/grpc/iosxr.py
Comment (samccann) [inline lib/ansible/plugins/grpc/iosxr.py]:
```suggestion
  - This gRPC plugin provides methods to connect and talk to Cisco IOS XR
```

File: test/integration/targets/grpc_nw_get/tests/iosxr/basic.yaml
Comment (samccann) [inline test/integration/targets/grpc_nw_get/tests/iosxr/basic.yaml]: not sure if this should be "wrong display type" ?

File: test/integration/targets/grpc_nw_get/tests/iosxr/basic.yaml
Comment (samccann) [inline test/integration/targets/grpc_nw_get/tests/iosxr/basic.yaml]: same question -= should it be wrong?

**Review by samccann — COMMENTED** [review]
A few nit docs comments.

I also created https://github.com/ansible/ansible/issues/52997 so we can add grpc into the platform options table. It requires a bit of doc cleanup so it all fits so I'll take that issue and request your review when I'm done.

File: lib/ansible/constants.py
Comment (ganeshrn) [reply lib/ansible/constants.py]: ansible-doc CI test fails if the plugin is not added as a configurable plugin that's the reason it is added here and basic configuration document is added in plugin

File: lib/ansible/plugins/grpc/__init__.py
Comment (ganeshrn) [reply lib/ansible/plugins/grpc/__init__.py:26]: The grpc protocol does not define any specific api's or RPC as in the case of netconf and restconf. The gRPC api's support for any platform is defined in the protobuf files, that's the reason these methods are not defined as `abstractmethod` as some of the platform may not have the need to implement these api's

File: lib/ansible/plugins/grpc/__init__.py
Comment (ganeshrn) [reply lib/ansible/plugins/grpc/__init__.py:35]: Same as above

File: lib/ansible/plugins/grpc/__init__.py
Comment (ganeshrn) [reply lib/ansible/plugins/grpc/__init__.py:44]: It is supported for iosxrv and will be added later on

**Comment by ganeshrn** [comment]
@Qalthos IMO gRPC plugin is required because the implementation varies based on the capabilities defined in the python bindings generated from protocol buffers and varies based on platform. Also, the protocol buffer is added in the individual plugin folder instead of `mdoule_utils` because the grpc modules does not directly consume these python bindings. IMO only the code that can be reused in modules should be placed under `module_utils`. Also the stub object (created form python binding) requires the gRPC channel object to pass to it https://github.com/ansible/ansible/pull/52857/files#diff-640ddcbe8e941027a8af1ebf0841390bR30 that's other for keeping the python bindings in grpc plugin folder.

**Review by samccann — APPROVED** [review]
docs portion LGTM

File: lib/ansible/plugins/grpc/__init__.py
Comment (Qalthos) [reply lib/ansible/plugins/grpc/__init__.py:26]: I guess if these methods aren't guaranteed to exist, then why present them in the base class at all? What is the benefit of having this method exist but be unimplemented on an implementation that doesn't use it?

**Comment by nitzmahone** [comment]
Uhh, adding yet more new plugin types is probably something that deserves at least a modicum of discussion with the core team, especially when there is only one implementation (how do you know you've got the abstraction correct with a single implemented use case?). This definitely needs some discussion before being merged; we're trying really hard to stem the tide of new plugin types and API surface areas that we'll have to support for many years, so this decision can't come lightly.

**Comment by mattclay** [comment]
Closing, now that network content has moved into collections.

**Pull request closed without merging**
